File: static/viewer/gemeinsam/stoffgrenze.js
/**
 * Stoffgrenze — hält den Weichgewebe-Zuschlag der Kleidung aus dem Körper.
 *
 * Die JavaScript-Fassung von `TheatreJS/ModelPhysik/stoffgrenze.py`, für
 * das Weichgewebe LIVE in der Szene (11.09.2026). Bis dahin bekam die
 * Kleidung im Browser die HALBE Stärke, weil die Grenze fehlte — mit voller
 * Stärke schob ein 25-mm-Zuschlag das T-Shirt mit 13 mm Hautabstand durch
 * die Haut. Dieselbe Rechnung wie in Python, je Punkt:
 *
 *     aussen     = <stoff − körper, n>          (positiv = außerhalb)
 *     spielraum  = max(aussen − MINDESTABSTAND, 0)
 *     nachInnen  = −<versatz, n>
 *     zuviel     → versatz += (nachInnen − spielraum) · n
 *
 * Nur der EINWÄRTS gerichtete Anteil wird gekürzt; ein Stück, das seitlich
 * mitschwingt, darf das weiter tun.
 *
 * EINE NÄHERUNG gegenüber Python, wegen der Bildrate: Welcher Körperpunkt
 * der nächste ist, wird EINMAL in Ruhe bestimmt (`Punktgitter`), nicht je
 * Bild neu. Der Stoff hängt über seine Hautgewichte ohnehin am nächsten
 * Dreieck — Punkt und Körperpunkt bewegen sich gemeinsam (Gleichlauf
 * 13 / 13 / 13 mm, `hbfilm.py`). Die NORMALE dagegen kommt je Bild aus dem
 * verformten Körper: Sie ist es, die beim Gehen ihre Richtung ändert.
 *
 * DIE RICHTUNG KOMMT AUS DEM KÖRPER, nie aus dem Stoff — die Flächennormale
 * eines Schnittteils hängt an seiner Wickelrichtung (viermal im Projekt
 * zugeschlagen, `stoffkorrektur.py`). Ob das Körpernetz nach außen
 * gewickelt ist, wird geprüft, nicht angenommen.
 */
import { Punktgitter } from './punktgitter.js';

export class Stoffgrenze {
    /** Wie in Python: der Kollisionsabstand der Drapierung (6 mm). */
    static MINDESTABSTAND = 0.006;
    /** Tiefer als das muss ein Punkt stecken, um als „im Körper" zu zählen. */
    static TOLERANZ = 0.001;

    /**
     * @param koerperRuhe    Float32Array nK*3, Körper in Ruhe (Netzraum)
     * @param koerperDreiecke Uint16/Uint32Array m*3
     * @param stoffRuhe      Float32Array nS*3, Stoff in Ruhe (derselbe Raum)
     */
    constructor(koerperRuhe, koerperDreiecke, stoffRuhe) {
        this.nK = koerperRuhe.length / 3;
        this.nS = stoffRuhe.length / 3;
        this.dreiecke = koerperDreiecke;
        this.naechster = new Punktgitter(koerperRuhe).alleNaechsten(stoffRuhe);
        this._flaechen();
        this.normalen = new Float64Array(this.benutzt.length * 3);
        this._normalen(koerperRuhe, null);
        this.aussen = this._aussen(koerperRuhe);
        this.gekuerzt = 0;
    }

    /** Die BENUTZTEN Körperpunkte und je Punkt seine Dreiecke (CSR). */
    _flaechen() {
        const platz = new Int32Array(this.nK).fill(-1);
        const benutzt = [];
        this.platz = new Int32Array(this.nS);
        for (let i = 0; i < this.nS; i++) {
            const j = this.naechster[i];
            if (j < 0) { this.platz[i] = -1; continue; }
            if (platz[j] < 0) { platz[j] = benutzt.length; benutzt.push(j); }
            this.platz[i] = platz[j];
        }
        this.benutzt = Int32Array.from(benutzt);
        const anzahl = new Int32Array(benutzt.length + 1);
        const d = this.dreiecke, m = d.length / 3;
        for (let t = 0; t < m; t++) {
            for (let e = 0; e < 3; e++) { const s = platz[d[3 * t + e]]; if (s >= 0) anzahl[s + 1] += 1; }
        }
        for (let s = 0; s < benutzt.length; s++) anzahl[s + 1] += anzahl[s];
        this.flStart = anzahl;
        this.flDreieck = new Int32Array(anzahl[benutzt.length]);
        const fuell = new Int32Array(benutzt.length);
        for (let t = 0; t < m; t++) {
            for (let e = 0; e < 3; e++) {
                const s = platz[d[3 * t + e]];
                if (s >= 0) { this.flDreieck[this.flStart[s] + fuell[s]] = t; fuell[s] += 1; }
            }
        }
    }

    /** Punktnormalen der benutzten Körperpunkte aus dem Körper DIESES Bildes. */
    _normalen(koerper, zuschlag) {
        const d = this.dreiecke, n = this.normalen;
        const px = (i) => koerper[3 * i] + (zuschlag ? zuschlag[3 * i] : 0);
        const py = (i) => koerper[3 * i + 1] + (zuschlag ? zuschlag[3 * i + 1] : 0);
        const pz = (i) => koerper[3 * i + 2] + (zuschlag ? zuschlag[3 * i + 2] : 0);
        for (let s = 0; s < this.benutzt.length; s++) {
            let nx = 0, ny = 0, nz = 0;
            for (let f = this.flStart[s]; f < this.flStart[s + 1]; f++) {
                const t = this.flDreieck[f];
                const a = d[3 * t], b = d[3 * t + 1], c = d[3 * t + 2];
                const ax = px(a), ay = py(a), az = pz(a);
                const bx = px(b) - ax, by = py(b) - ay, bz = pz(b) - az;
                const cx = px(c) - ax, cy = py(c) - ay, cz = pz(c) - az;
                nx += by * cz - bz * cy; ny += bz * cx - bx * cz; nz += bx * cy - by * cx;
            }
            const l = Math.hypot(nx, ny, nz) || 1e-12;
            n[3 * s] = nx / l; n[3 * s + 1] = ny / l; n[3 * s + 2] = nz / l;
        }
    }

    /** +1 oder −1: zeigen die Normalen nach außen? Mehrheit gegen die Mitte. */
    _aussen(koerper) {
        let mx = 0, my = 0, mz = 0;
        for (let i = 0; i < this.nK; i++) { mx += koerper[3 * i]; my += koerper[3 * i + 1]; mz += koerper[3 * i + 2]; }
        mx /= this.nK; my /= this.nK; mz /= this.nK;
        let dafuer = 0;
        for (let s = 0; s < this.benutzt.length; s++) {
            const j = this.benutzt[s];
            const skalar = (koerper[3 * j] - mx) * this.normalen[3 * s]
                + (koerper[3 * j + 1] - my) * this.normalen[3 * s + 1]
                + (koerper[3 * j + 2] - mz) * this.normalen[3 * s + 2];
            if (skalar > 0) dafuer += 1;
        }
        return dafuer * 2 >= this.benutzt.length ? 1 : -1;
    }

    /**
     * Den Versatz IN PLACE kürzen. `stoff` sind die LBS-Punkte des Stoffs
     * (vor dem Zuschlag), `koerper` die LBS-Punkte des Körpers und
     * `zuschlag` dessen Zuschlag desselben Bildes (oder null).
     * Gibt die Zahl der gekürzten Punkte zurück.
     */
    kuerzen(stoff, versatz, koerper, zuschlag) {
        this._normalen(koerper, zuschlag);
        const n = this.normalen, MIN = Stoffgrenze.MINDESTABSTAND, vz = this.aussen;
        let zahl = 0;
        for (let i = 0; i < this.nS; i++) {
            const s = this.platz[i];
            if (s < 0) continue;
            const j = this.benutzt[s];
            const nx = vz * n[3 * s], ny = vz * n[3 * s + 1], nz = vz * n[3 * s + 2];
            const kx = koerper[3 * j] + (zuschlag ? zuschlag[3 * j] : 0);
            const ky = koerper[3 * j + 1] + (zuschlag ? zuschlag[3 * j + 1] : 0);
            const kz = koerper[3 * j + 2] + (zuschlag ? zuschlag[3 * j + 2] : 0);
            const aussen = (stoff[3 * i] - kx) * nx + (stoff[3 * i + 1] - ky) * ny + (stoff[3 * i + 2] - kz) * nz;
            const spielraum = Math.max(aussen - MIN, 0);
            const nachInnen = -(versatz[3 * i] * nx + versatz[3 * i + 1] * ny + versatz[3 * i + 2] * nz);
            if (nachInnen > spielraum) {
                const u = nachInnen - spielraum;
                versatz[3 * i] += u * nx; versatz[3 * i + 1] += u * ny; versatz[3 * i + 2] += u * nz;
                zahl += 1;
            }
        }
        this.gekuerzt = zahl;
        return zahl;
    }

    /**
     * Anteil der Stoffpunkte IM Körper (Prozent) und die größte Tiefe (mm),
     * mit den Normalen des letzten `kuerzen`-Aufrufs bzw. der Ruhe.
     */
    durchdringung(stoff, koerper, zuschlag) {
        const n = this.normalen, vz = this.aussen;
        let drin = 0, tiefste = 0, gezaehlt = 0;
        for (let i = 0; i < this.nS; i++) {
            const s = this.platz[i];
            if (s < 0) continue;
            const j = this.benutzt[s];
            gezaehlt += 1;
            const kx = koerper[3 * j] + (zuschlag ? zuschlag[3 * j] : 0);
            const ky = koerper[3 * j + 1] + (zuschlag ? zuschlag[3 * j + 1] : 0);
            const kz = koerper[3 * j + 2] + (zuschlag ? zuschlag[3 * j + 2] : 0);
            const aussen = vz * ((stoff[3 * i] - kx) * n[3 * s] + (stoff[3 * i + 1] - ky) * n[3 * s + 1]
                + (stoff[3 * i + 2] - kz) * n[3 * s + 2]);
            if (aussen < -Stoffgrenze.TOLERANZ) { drin += 1; if (-aussen > tiefste) tiefste = -aussen; }
        }
        return { prozent: gezaehlt ? 100 * drin / gezaehlt : 0, tiefe_mm: tiefste * 1000 };
    }
}

File: TheatreJS/ModelPhysik/stoffgrenze.py
# -*- coding: utf-8 -*-
u"""Haelt den Physik-Zuschlag aus dem Koerper heraus.

DER BEFUND, DER DAZU GEFUEHRT HAT (11.09.2026): Velocity Skinning ist fuer
KOERPER entworfen — dort ist jeder Punkt an Knochen gebunden, und ein
Zuschlag von 20 mm bleibt Gewebe. Ein Kleidungsstueck haengt dagegen lose,
und sein Abstand zur Haut ist gemessen 13 mm (T-Shirt) bzw. 27 mm (Hose).
Ein Zuschlag von 28 mm schiebt es glatt durch den Koerper: im Bild klaffen
Loecher, durch die die Haut sieht.

DIE GRENZE STEHT NICHT IM ORIGINAL, und das ist kein Versehen des
Upstream — er animiert keine Kleidung. Sie ist auch keine Simulation:
Der Zuschlag wird je Punkt so weit gekuerzt, dass ein Mindestabstand zur
Haut bleibt. Nach aussen bleibt er ungekuerzt; ein Stueck, das vom Koerper
wegschwingt, soll das duerfen.

Gemessen wird gegen den KOERPER DESSELBEN BILDES, nicht gegen die
Ruhelage — beim Gehen wandert die Haut unter dem Stoff.

DER KOERPER KOMMT AUCH VON INNEN (Edgar, 11.09.2026, mit Bild: „bei einer
animation kommt der Koerper durch die Kleidung hindurch"). Die erste
Fassung kuerzte nur den EINWAERTS gerichteten Anteil des Stoffzuschlags;
bekam der Koerper selbst einen groesseren Zuschlag als der Stoff darueber,
stand die Haut durch — bei 13 mm Hautabstand (T-Shirt) nie sichtbar, bei
einer Leggings auf 2 mm sofort. Gemessen in der Szene (Dance1, Sprung aus
der T-Pose): Koerperzuschlag 188 mm, Stoff 109 mm, 6,3 % der Leggings-
punkte im Koerper. Deshalb gilt jetzt ein SOLLABSTAND je Punkt: sein
Ruheabstand, hoechstens `MINDESTABSTAND` — und wer darunter liegt, wird
auf ihn gehoben, gleich ob der Stoff nach innen oder der Koerper nach
aussen gegangen ist.
"""
import numpy as np


class Stoffgrenze:
    u"""Kuerzt Verschiebungen, die in den Koerper hineinzeigen."""

    #: So nah darf der Stoff der Haut kommen. Der Kollisionsabstand der
    #: Drapierung liegt bei 6 mm (CLAUDE.md, 08.09.2026, nach dem Befund
    #: „Haut scheint durch die Hose"); darunter zu gehen hiesse, den dort
    #: teuer erarbeiteten Abstand wieder herzugeben.
    MINDESTABSTAND = 0.006

    def __init__(self, koerperpunkte, koerperdreiecke):
        from scipy.spatial import cKDTree
        self.punkte = np.asarray(koerperpunkte, dtype=np.float64)
        self.dreiecke = np.asarray(koerperdreiecke, dtype=np.int64)
        self.baum = cKDTree(self.punkte)
        self.normalen = self._normalen()

    def _normalen(self):
        u"""Punktnormalen des Koerpers, aus den Flaechennormalen gemittelt.

        DIE RICHTUNG MUSS AUS DEM KOERPER KOMMEN, nie aus dem Stoff. Die
        Flaechennormale eines Schnittteils haengt an seiner Wickelrichtung
        und kann zum Koerper zeigen — dieselbe Falle hat im Projekt schon
        viermal zugeschlagen (`stoffkorrektur.py`, 06.09.2026).
        """
        a = self.punkte[self.dreiecke[:, 0]]
        b = self.punkte[self.dreiecke[:, 1]]
        c = self.punkte[self.dreiecke[:, 2]]
        flaeche = np.cross(b - a, c - a)
        aus = np.zeros_like(self.punkte)
        for spalte in range(3):
            np.add.at(aus, self.dreiecke[:, spalte], flaeche)
        laengen = np.linalg.norm(aus, axis=1, keepdims=True)
        aus = aus / np.maximum(laengen, 1e-12)
        return aus * self._aussen(aus)

    def _aussen(self, normalen):
        u"""+1 oder -1 — zeigen die Normalen nach aussen?

        Ueber das SIGNIERTE VOLUMEN, nicht die Mehrheit gegen den
        Schwerpunkt: Die Grenze wird je Bild aus dem POSIERTEN Koerper
        gebaut, und die Mehrheit kippte in Dance1 bei 32 % (11.09.2026,
        feines 70K-Netz) — der Stoff wurde in den Koerper GEZOGEN, im Bild
        lag die Leggings innen. Das Volumen haengt an der Wicklung, und die
        aendert keine Pose. `normalen` bleibt in der Signatur, damit die
        Browser-Fassung (`stoffgrenze.js`) denselben Aufruf traegt.
        """
        a = self.punkte[self.dreiecke[:, 0]]
        b = self.punkte[self.dreiecke[:, 1]]
        c = self.punkte[self.dreiecke[:, 2]]
        vol = float(np.einsum('ij,ij->i', a, np.cross(b, c)).sum())
        return 1.0 if vol >= 0 else -1.0

    def abstand(self, stoffpunkte):
        u"""Der vorzeichenbehaftete Abstand je Stoffpunkt (positiv = aussen)."""
        _abstand, naechster = self.baum.query(stoffpunkte)
        rest = stoffpunkte - self.punkte[naechster]
        return np.sum(rest * self.normalen[naechster], axis=1)

    def sollabstand(self, stoffruhe):
        u"""Je Stoffpunkt: sein Ruheabstand, hoechstens `MINDESTABSTAND`.

        Auf dieser Grenze (aus der RUHELAGE gerechnet) haelt `kuerzen`
        jeden Punkt — eine Leggings auf 2 mm bleibt auf 2 mm, ein T-Shirt
        auf 13 mm wird nicht naeher als 6 mm gelassen.
        """
        return np.minimum(self.abstand(stoffruhe), self.MINDESTABSTAND)

    def kuerzen(self, stoffpunkte, versatz, soll=None):
        u"""Der erlaubte Teil des Zuschlags, je Punkt.

        `soll` ist der Sollabstand je Punkt (`sollabstand`); ohne ihn gilt
        `MINDESTABSTAND` fuer alle. Rueckgabe ist der gekuerzte Versatz und
        die Zahl der Punkte, bei denen gekuerzt wurde.
        """
        _abstand, naechster = self.baum.query(stoffpunkte)
        normale = self.normalen[naechster]
        rest = stoffpunkte - self.punkte[naechster]
        # Der vorzeichenbehaftete Abstand: positiv = ausserhalb.
        aussen = np.sum(rest * normale, axis=1)
        nach_innen = -np.sum(versatz * normale, axis=1)
        # Wo der Punkt MIT Versatz laege, gegen das, was ihm zusteht.
        ist = aussen - nach_innen
        ziel = self.MINDESTABSTAND if soll is None else np.asarray(soll)
        zuviel = ist < ziel
        if not zuviel.any():
            return versatz, 0
        gekuerzt = np.array(versatz)
        # Nur ENTLANG DER NORMALE anheben, den Rest lassen: Ein Stueck, das
        # seitlich mitschwingt, soll das weiter tun.
        ueberschuss = (ziel - ist)[zuviel]
        gekuerzt[zuviel] += ueberschuss[:, None] * normale[zuviel]
        return gekuerzt, int(zuviel.sum())

    #: Tiefer als das muss ein Punkt stecken, um zu zaehlen. Ohne Toleranz
    #: meldete der Messer am T-Shirt 6,09 % „im Koerper" — bei einem
    #: tiefsten Punkt von 0,1 mm. Das sind Punkte, die AUF der Haut
    #: liegen und je nach Normale um Bruchteile eines Millimeters
    #: innen landen; sichtbar ist davon nichts.
    TOLERANZ = 0.001

    def durchdringung(self, stoffpunkte):
        u"""Anteil der Stoffpunkte IM Koerper, und die groesste Tiefe.

        Die Probe, die zaehlt. Sie fragt je STOFFpunkt, ob er im Koerper
        steckt — nicht je Koerperpunkt, ob Stoff dahinter liegt: Der
        Koerper ist geschlossen, „innen" ist dort wohldefiniert, und
        offene Saeume spielen keine Rolle (CLAUDE.md, 07.09.2026, nachdem
        der Messer viermal falsch herum stand).
        """
        _abstand, naechster = self.baum.query(stoffpunkte)
        rest = stoffpunkte - self.punkte[naechster]
        aussen = np.sum(rest * self.normalen[naechster], axis=1)
        drin = aussen < -self.TOLERANZ
        # `aussen` ist innen NEGATIV — der tiefste Punkt ist das MINIMUM.
        # Bis zum 11.09.2026 stand hier `.max()`, also der flachste Punkt,
        # als „groesste Tiefe" in jeder Bilanz. Gefunden hat es die
        # JavaScript-Gegenprobe (`test_js_stoffgrenze.py`): 19,0 mm gegen
        # 1,0 mm bei denselben Punkten.
        return (float(drin.mean()) * 100.0,
                float(-aussen[drin].min()) * 1000.0 if drin.any() else 0.0)
